compiler/rustc_trait_selection/src/traits/select/confirmation.rs

```diff
@@ -693,12 +694,28 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
 
         debug!(?obligation, ?generator_def_id, ?substs, "confirm_generator_candidate");
 
-        let trait_ref = self.generator_trait_ref_unnormalized(obligation, substs);
+        let trait_ref = if is_async {
+            // FIXME(swatinem): pretty much just copy-pasted generator_trait_ref_unnormalized
+            // and generator_trait_ref_and_outputs:
+            let gen_sig = substs.as_generator().poly_sig();
+
+            use rustc_middle::ty::TypeVisitable;
+            debug_assert!(!self_ty.has_escaping_bound_vars());
+            let trait_ref = ty::TraitRef {
+                def_id: obligation.predicate.def_id(),
+                substs: self.tcx().mk_substs_trait(self_ty, &[]),
+            };
+            gen_sig
+                .map_bound(|sig| (trait_ref, sig.return_ty))
+                .map_bound(|(trait_ref, ..)| trait_ref)
+        } else {
+            self.generator_trait_ref_unnormalized(obligation, substs)
```

Comment:
`generator_trait_ref_unnormalized` is quite small, it could probably be inlined here.